services/alertmanager/alertmanager.go

```diff
@@ -60,120 +66,262 @@ const (
 
 // Service is responsible for interactions with Alertmanager.
 type Service struct {
-	db *reform.DB
-	l  *logrus.Entry
+	db     *reform.DB
+	client *http.Client
+
+	l        *logrus.Entry
+	reloadCh chan struct{}
 }
 
 // New creates new service.
 func New(db *reform.DB) *Service {
 	return &Service{
-		db: db,
-		l:  logrus.WithField("component", "alertmanager"),
+		db:       db,
+		client:   new(http.Client), // TODO instrument with utils/irt; see vmalert package
+		l:        logrus.WithField("component", "alertmanager"),
+		reloadCh: make(chan struct{}, 1),
 	}
 }
 
 // Run runs Alertmanager configuration update loop until ctx is canceled.
 func (svc *Service) Run(ctx context.Context) {
+	// If you change this and related methods,
+	// please do similar changes in victoriametrics and vmalert packages.
+
 	svc.l.Info("Starting...")
 	defer svc.l.Info("Done.")
 
-	err := dir.CreateDataDir(alertmanagerDir, "pmm", "pmm", dirPerm)
-	if err != nil {
+	if err := dir.CreateDataDir(alertmanagerDir, "pmm", "pmm", dirPerm); err != nil {
 		svc.l.Error(err)
 	}
-	err = dir.CreateDataDir(alertmanagerDataDir, "pmm", "pmm", dirPerm)
-	if err != nil {
+	if err := dir.CreateDataDir(alertmanagerDataDir, "pmm", "pmm", dirPerm); err != nil {
+		svc.l.Error(err)
+	}
+
+	if err := svc.generateBaseConfig(); err != nil {
 		svc.l.Error(err)
 	}
 
-	svc.generateBaseConfig()
-	svc.updateConfiguration(ctx)
+	// reloadCh, configuration update loop, and RequestConfigurationUpdate method ensure that configuration
+	// is reloaded when requested, but several requests are batched together to avoid too often reloads.
+	// That allows the caller to just call RequestConfigurationUpdate when it seems fit.
+	if cap(svc.reloadCh) != 1 {
+		panic("reloadCh should have capacity 1")
+	}
 
-	// we don't have "configuration update loop" yet, so do nothing
-	// TODO implement loop similar to victoriametrics.Service.Run
+	for {
+		select {
+		case <-ctx.Done():
+			return
 
-	<-ctx.Done()
-}
+		case <-svc.reloadCh:
+			// batch several update requests together by delaying the first one
+			sleepCtx, sleepCancel := context.WithTimeout(ctx, updateBatchDelay)
+			<-sleepCtx.Done()
+			sleepCancel()
```

Comment:
How that is supposed to work with buffered channel? When I hit `ChangeSettings` twice:
1) first message sent to the  channel
2) first message read from channel and reading goroutine stoped for batch delay
3) second message is sent to the channel
4) after batch delay first message processed
5) second message is read form channel
6) ....

if we will change buffered channel to unbuffered second message will be not sent in channel because write is done with select.

Am I missing something?

Reply:
If channel will be unbuffered, the following scenario is possible:

1. first message set to the channel
2. first message read from channel and reading goroutine stoped for batch delay
3. after batch delay first message processed
4. while first message is processed, second message is being sent to the channel – and dropped, because Run is not reading from the channel at that moment

Reply:
Then whole that mechanism is broken and should be done a different way (obvoiusly not now). 

We can spawn new goroutine for message processing, or we can accept that extra time spent on first message processin as addittion to the batch delay.

Ok, it is not the proper time for that.